launchlib/launcher.go

```diff
@@ -95,13 +95,24 @@ func compileCmdFromConfig(
 			staticConfig.JavaConfig.Classpath))
 		_, _ = fmt.Fprintf(logger, "Classpath: %s\n", classpath)
 
+		var jvmOpts []string
+
+		if isEnvVarSet("CONTAINER") && staticConfig.ContainerSupport {
+			_, _ = fmt.Fprintln(logger, "Container support enabled")
+			jvmOpts = append(jvmOpts, filterHeapArgs(staticConfig.JavaConfig.JvmOpts)...)
+			jvmOpts = append(jvmOpts, filterHeapArgs(customConfig.JvmOpts)...)
+			jvmOpts = append(jvmOpts, []string{"-XX:+UseContainerSupport", "-XX:InitialRAMPercentage=80.0", "-XX:MaxRAMPercentage=80.0"}...)
```

Comment:
In that case we wouldn't need additional config, if CONTAINER is set, we strip `MaxRAM`, however current services aren't impacted by that, and behavior isn't modified.

Reply:
I haven't seen MaxRAM before: so you're suggesting parsing Xmx to figure out bytes to pass to MaxRAM, WHEN containerSupport isn't set?

Reply:
I have read some random people on the internet suggesting MaxRAMPercentage is a safer bet: https://medium.com/adorsys/jvm-memory-settings-in-a-container-environment-64b0840e1d9e

Reply:
From what I'm reading, if you're only setting MaxRAM, it will use MaxRAMFraction to actually figure out the heap size, which by default is 4, which means you're giving yourself 25% of memory available in the container. So we'd need to adjust that anyway? I personally prefer to read percentages than 1/MaxRAMFraction to figure out what it's going to give to the process.

Reply:
Idea is:

Excavate out a change to replace `-Xms` and `-Xmx` with `-XX:MaxRAM=<1.25*Xmx> -XX:MaxRAMPercentage=80 -XX:InitialRAMPercentage=80*(original_xms / original_xmx)` 

Then, all we have to do in the launcher is filter out `-XX:MaxRAM` when the `CONTAINER` envvar is present, we don't have to worry about xmx and xms at all.

Reply:
Note the numbers are based on the assumption that heap should be 80% of total ram, that may be inaccurate in which case we'd need to update values accordingly.

Reply:
That sounds like a lot more work to me, for not much gain? what's wrong with MaxRAMPercentage? (I am definitely assuming that it actually works, but to do that I will actually have to test it out, so it might be that your way is the only way)

Reply:
In general, I think the assumption we're all making is that in general we'll be running with pretty small heaps, so sacrificing 20% does not feel bad. But for something like 8GB heap, assuming that native memory grows linearly is maybe not an always bullet-proof assumption? but it's probably good enough to work in most cases.

I feel we should probably have a metric that allows us to see actual heap usage somehow

Reply:
Filtering out `MaxRAM` is easier imo than replacing xms and xmx with another pair of parameters, I like the idea of keeping this project as simple as possible (remove exactly one parameter, if it exists). I like the idea of reusing as much configuration as possible in non-containerized environments as well.

> I think the assumption we're all making is that in general we'll be running with pretty small heaps, so sacrificing 20% does not feel bad. But for something like 8GB heap, assuming that native memory grows linearly is maybe not an always bullet-proof assumption? but it's probably good enough to work in most cases.

We need to worry about nonheap memory as well: thread overhead, direct memory (ByteBuffer.allocateDirect, defaults to a maximum equal to your heap), native jni overhead, metaspace, etc. Some fun detail here: https://shipilev.net/jvm/anatomy-quarks/12-native-memory-tracking/

> I feel we should probably have a metric that allows us to see actual heap usage somehow

Our memory pool metrics from tritium give us this data, but not in the easiest to consume formats (have to sum the pool sizes).

Reply:
Ok do my final thing is that: even if we excavate it out, you’d still have launcher-custom, which you cannot excavate?

Reply:
Yep, if you explicitly set custom heap, that's still used. I think that's okay and more obvious.

Reply:
The more I think about it, the more I feel like we should have different distributions: classic tar, container and container with native image. And it shouldn’t be the go-Java-launcher that deals with this complexity, but it’s dealt with at build time.

Reply:
The gradle distribution plugin supports multiple distributions nicely: https://docs.gradle.org/current/userguide/distribution_plugin.html

Reply:
> I think that's okay and more obvious.

I think the problem with this is that ALL products have this overriden, because that's how you set heap right now. Most products just map launcherCustom.heap to the launcher-custom.yml file and call it a day. Virtually no products actually use "staticLauncher" from sls-packaging to set their default heap sizes.

Reply:
re multiple distribution types based on container/non-container: I think this runs into an issue with our config/manifest distribution, which is expected to match the configuraiton for a distribution with the same coordinates. This makes it difficult to support divergent configuration schemes based on the deployment model. I don't know the specifics though, and I may be incorrect about how it works.

Reply:
After much deliberating, we have settled on this being the best solution to the problem, where go-java-launcher is reponsible for unsetting -Xmx and -Xms, and setting MaxRamPercentage/InitialRAMPercentage if both not set.